### ISSUE TITLE
Incorrect backward pass for non-contiguous output grads

### ISSUE THREAD

**Issue by jlamypoirier** [Dao-AILab/flash-attention#355]
I'm trying to update Megatron to flash attention 2 (https://github.com/bigcode-project/Megatron-LM/pull/72), but I'm getting incorrect gradients. It seems to be caused by output gradients being non-contiguous. That would be fixable with a `contiguous` call, though a in-pace handling would be more efficient (if doable)

I tested by modifying `FlashAttnFunc.backward`:
```
    @staticmethod
    def backward(ctx, dout, *args):
        q, k, v, out, softmax_lse, rng_state = ctx.saved_tensors
        #dout=dout.contiguous()
        debug_print("Q", q, level=6)
        debug_print("K", k, level=6)
        debug_print("V", v, level=6)
        debug_print("out", out, level=6)
        debug_print("softmax_lse", softmax_lse, level=6)
        #debug_print("rng_state", rng_state, level=6)
        if rng_state is not None:
            cur_rng_state = torch.cuda.get_rng_state()
            torch.cuda.set_rng_state(rng_state)
        dq, dk, dv = torch.empty_like(q), torch.empty_like(k), torch.empty_like(v)
        _flash_attn_backward(
            dout, q, k, v, out, softmax_lse,
            dq, dk, dv, ctx.dropout_p, ctx.softmax_scale, ctx.causal
        )
        dq = dq[..., :dout.shape[-1]]  # We could have padded the head dimension
        dk = dk[..., :dout.shape[-1]]
        dv = dv[..., :dout.shape[-1]]
        if rng_state is not None:
            torch.cuda.set_rng_state(cur_rng_state)
        debug_print("GRAD ATTN OUTPUT", dout, level=6)
        debug_print("GRAD Q", dq, level=6)
        debug_print("GRAD K", dk, level=6)
        debug_print("GRAD V", dv, level=6)
        return dq, dk, dv, None, None, None, None, None, None, None, None, None, None, None
```

Which gives the incorrect output (different from disabled flash attn)
```
Q, name=None, shape=[2, 32, 4, 128], dtype=torch.float16, device=cuda:0, stats=(-8.365e-04, 0.4958), stride=(16384, 512, 128, 1), storage=140041285861376, storage size=65536, storage stats=(-8.365e-04, 0.4958), samples (step=8191) = ['0.1348', '0.3953', '0.1809', '-0.7539']
K, name=None, shape=[2, 32, 1, 128], dtype=torch.float16, device=cuda:0, stats=(3.624e-03, 0.4957), stride=(4096, 128, 128, 1), storage=140041286057984, storage size=16384, storage stats=(3.624e-03, 0.4957), samples (step=1) = ['-0.4619', '0.0524', '2.766e-03', '0.4041']
V, name=None, shape=[2, 32, 1, 128], dtype=torch.float16, device=cuda:0, stats=(0.0146, 0.5012), stride=(4096, 128, 128, 1), storage=140041286074368, storage size=16384, storage stats=(0.0146, 0.5012), samples (step=1) = ['-0.2817', '-0.2369', '0.4480', '0.3354']
out, name=None, shape=[2, 32, 4, 128], dtype=torch.float16, device=cuda:0, stats=(0.0126, 0.2003), stride=(16384, 512, 128, 1), storage=140041286090752, storage size=65536, storage stats=(0.0126, 0.2003), samples (step=8191) = ['-0.2817', '0.0939', '-0.0957', '0.1498']
softmax_lse, name=None, shape=[2, 4, 32], dtype=torch.float32, device=cuda:0, stats=(2.5709, 0.8527), stride=(128, 32, 1), storage=140042307530752, storage size=1024, storage stats=(2.5709, 0.8527), samples (step=63) = ['-0.2922', '3.5131', '3.5108', '3.3868']
GRAD ATTN OUTPUT, name=None, shape=[2, 32, 4, 128], dtype=torch.float16, device=cuda:0, stats=(0.9361, 128.7211), stride=(512, 1024, 128, 1), storage=140042306595328, storage size=65536, storage stats=(0.9361, 128.7211), samples (step=8191) = ['-12.0234', '-117.0000', '-325.7500', '-16.9062']
GRAD Q, name=None, shape=[2, 32, 4, 128], dtype=torch.float16, device=cuda:0, stats=(-7.788e-03, 9.2473), stride=(16384, 512, 128, 1), storage=140041285926912, storage size=65536, storage stats=(-7.788e-03, 9.2473), samples (step=8191) = ['-1.431e-06', '5.8633', '2.2891', '2.6152']
GRAD K, name=None, shape=[2, 32, 1, 128], dtype=torch.float16, device=cuda:0, stats=(-0.0250, 18.7403), stride=(4096, 128, 128, 1), storage=140041286287360, storage size=16384, storage stats=(-0.0250, 18.7403), samples (step=1) = ['22.4688', '60.2188', '-71.6875', '-16.1406']
GRAD V, name=None, shape=[2, 32, 1, 128], dtype=torch.float16, device=cuda:0, stats=(3.7442, 88.2849), stride=(4096, 128, 128, 1), storage=140041286303744, storage size=16384, storage stats=(3.7442, 88.2849), samples (step=1) = ['226.1250', '27.6562', '-157.3750', '24.7500']
```

When I uncomment the `dout=dout.contiguous()` however I get the correct output
```
Q, name=None, shape=[2, 32, 4, 128], dtype=torch.float16, device=cuda:0, stats=(-8.365e-04, 0.4958), stride=(16384, 512, 128, 1), storage=140061150085120, storage size=65536, storage stats=(-8.365e-04, 0.4958), samples (step=8191) = ['0.1348', '0.3953', '0.1809', '-0.7539']
K, name=None, shape=[2, 32, 1, 128], dtype=torch.float16, device=cuda:0, stats=(3.624e-03, 0.4957), stride=(4096, 128, 128, 1), storage=140061150281728, storage size=16384, storage stats=(3.624e-03, 0.4957), samples (step=1) = ['-0.4619', '0.0524', '2.766e-03', '0.4041']
V, name=None, shape=[2, 32, 1, 128], dtype=torch.float16, device=cuda:0, stats=(0.0146, 0.5012), stride=(4096, 128, 128, 1), storage=140061150298112, storage size=16384, storage stats=(0.0146, 0.5012), samples (step=1) = ['-0.2817', '-0.2369', '0.4480', '0.3354']
out, name=None, shape=[2, 32, 4, 128], dtype=torch.float16, device=cuda:0, stats=(0.0126, 0.2003), stride=(16384, 512, 128, 1), storage=140061150314496, storage size=65536, storage stats=(0.0126, 0.2003), samples (step=8191) = ['-0.2817', '0.0939', '-0.0957', '0.1498']
softmax_lse, name=None, shape=[2, 4, 32], dtype=torch.float32, device=cuda:0, stats=(2.5709, 0.8527), stride=(128, 32, 1), storage=140062171754496, storage size=1024, storage stats=(2.5709, 0.8527), samples (step=63) = ['-0.2922', '3.5131', '3.5108', '3.3868']
GRAD ATTN OUTPUT, name=None, shape=[2, 32, 4, 128], dtype=torch.float16, device=cuda:0, stats=(0.9361, 128.7211), stride=(16384, 512, 128, 1), storage=140061150150656, storage size=65536, storage stats=(0.9361, 128.7211), samples (step=8191) = ['-12.0234', '-117.0000', '-325.7500', '-16.9062']
GRAD Q, name=None, shape=[2, 32, 4, 128], dtype=torch.float16, device=cuda:0, stats=(1.373e-03, 8.3627), stride=(16384, 512, 128, 1), storage=140061150380032, storage size=65536, storage stats=(1.373e-03, 8.3627), samples (step=8191) = ['-1.431e-06', '0.5962', '4.0273', '1.8525']
GRAD K, name=None, shape=[2, 32, 1, 128], dtype=torch.float16, device=cuda:0, stats=(-4.134e-05, 16.8225), stride=(4096, 128, 128, 1), storage=140061150511104, storage size=16384, storage stats=(-4.134e-05, 16.8225), samples (step=1) = ['26.5312', '44.8750', '-69.4375', '-23.0625']
GRAD V, name=None, shape=[2, 32, 1, 128], dtype=torch.float16, device=cuda:0, stats=(3.7442, 88.2849), stride=(4096, 128, 128, 1), storage=140061150527488, storage size=16384, storage stats=(3.7442, 88.2849), samples (step=1) = ['226.1250', '27.6562', '-157.3750', '24.7500']
```

(The debug print is defined as follow)
```

def format_number(x, prec=4, exp_threshold=3):
    digits=0 if x==0 else math.log10(abs(x))
    if math.isfinite(digits) and -exp_threshold<math.floor(digits)<prec+exp_threshold:
        return f"{x:.{prec}f}"
    else:
        return f"{x:.{prec-1}e}"


def debug_print(t, v, scale=1.0, level=2):
    if level<1:
        return
    # Not using getattr sometimes gives another tensor's name
    txt = f"{t}, name={v.param_name if hasattr(v, 'param_name') else None}," \
          f" shape={list(v.shape)}, dtype={v.dtype}, device={v.device}"
    if level>=2 and v.device.type != "meta":
        v_float = v.float()
        txt+=(
            f", stats=({format_number(v_float.mean().item() * scale)}"
            f", {format_number(v_float.std().item() * scale)})"
            f", stride={v.stride()}"
        )
        if level>=3:
            storage = v.untyped_storage()
            storage_float = torch.tensor(storage, dtype=v.dtype, device=v.device).float()
            txt += (
                f", storage={storage.data_ptr()}"
                f", storage size={storage.size()}"
                f", storage stats=({format_number(storage_float.mean().item() * scale)}"
                f", {format_number(storage_float.std().item() * scale)})"
            )
        if level>=4:
            target_samples=2**(level-4)
            shape=tuple(v.shape)
            step=max(v.numel()//target_samples, 1)
            while step>1 and any(step%s==0 for s in shape):
                step-=1
            samples=[format_number(x) for x in v.flatten()[:target_samples*step:step].tolist()]
            txt+=f", samples (step={step}) = {samples}"
    print(txt)
```

**Comment by jlamypoirier** [comment]
I think so. Could this relate to #338? I have been getting lots of these illegal memory accesses.

**Comment by tridao** [comment]
Could very much be related. I'm trying to debug these issues w the backward pass.

**Comment by jlamypoirier** [comment]
Already fixed

**Comment by jlamypoirier** [comment]
Actually not, the commit only checks for stride(-1) but the bug happens for non-contiguous dimensions 0 and 1.

**Comment by tridao** [comment]
That's good to know, let me take a closer look. Thanks for the bug report, this is very valuable!

**Comment by tridao** [comment]
I just push a commit to hopefully fix this. @jlamypoirier can you try it? If it works I'll cut a release.

**Comment by jlamypoirier** [comment]
Thanks for the quick fix! It seems to be working, at least for the few inputs I tried.

**Comment by jlamypoirier** [comment]
Fixed
